### ISSUE TITLE
[device without fully binderized hals] Library load issue on P

### ISSUE THREAD

**Issue by simonsmh** [phhusson/treble_experimentations#165]
HI, I tried the AOSP 9.0 GSIs and I meet problems that don't exist on AOSP 8.1 GSIs.
Due to vndk enforcement, the linker could not load libs under /system
Reference: https://source.android.com/devices/architecture/vndk/linker-namespace

It leads to camera failure on my device. I tried to fix the missing libs, but the process is endless. 
My device vendor runs under 8.1, it uses vndk-27, but the default path doesn't contain system/lib .

Part of AOSP 8 ld.config.txt
```
namespace.default.isolated = false
namespace.default.search.paths = /vendor/${LIB}/hw:/vendor/${LIB}/egl:/vendor/${LIB}:/system/${LIB}/vndk:/vendor/${LIB}/vndk-sp:/system/${LIB}/vndk-sp:/system/${LIB}
```

Part of AOSP 9 ld.config.27.txt
```
namespace.default.isolated = true
namespace.default.search.paths = /vendor/${LIB}/hw:/vendor/${LIB}/egl:/vendor/${LIB}:/vendor/${LIB}/vndk:/system/${LIB}/vndk-27:/vendor/${LIB}/vndk-sp:/system/${LIB}/vndk-sp-27
```

[Linker source code reference](https://github.com/LineageOS/android_bionic/blob/lineage-16.0/linker/linker.cpp)

[Two file differences](https://gist.github.com/simonsmh/efd410625a90ec25d4ac723965ed72e9/revisions)

Could we make it a patch to fix it?

-------
Also, after this fix, I meet another issue that break the camera:

```
08-31 11:26:31.393  2218  2218 E vndksupport: Could not load /vendor/lib/hw/camera.msm8996.so from default namespace: dlopen failed: cannot locate symbol "_ZN7android12IUidObserver10descriptorE" referenced by "/system/lib/libcameraservice.so"....
08-31 11:26:31.393  2218  2218 E HAL     : load: module=/vendor/lib/hw/camera.msm8996.so
08-31 11:26:31.393  2218  2218 E HAL     : unknown
```
How could I fix this issue?

**Comment by simonsmh** [comment]
The source code is under https://github.com/OP3Treble

**Comment by phhusson** [comment]
Well, don't change the ld.config, and copy the libs from an aosp 8.1 until it stops complaining about missing libs.

**Comment by simonsmh** [comment]
From 8.1? is that working with aosp 9.0?

**Comment by phhusson** [comment]
Yes that's how vndk works, it gives compatibility by simply having litteraly the same libs

**Comment by simonsmh** [comment]
Thanks! Let me try it.

**Comment by simonsmh** [comment]
Thank you! After replacing libcameraservice.so, problem solved.

**Comment by phhusson** [comment]
I see that you still have ugly https://github.com/OP3Treble/linkerfix
FYI you can have a clean fix by doing `patchelf --add-needed libcamera_shim.so /vendor/lib/hw/camera.msm8996.so`, without requiring any change to linker

**Comment by simonsmh** [comment]
Actually it's not necessary since we have got vendor update. The shim is not necessary on ly device. It's not a problem anymore.
However, there is another problem in R linker namespace. When I tried to move forward to R GSIs, I found that It's no longer possible to make `/vendor/lib` libs to load `/system/lib` libs, which I used to use a prebuilt `/vendor/lib/libcameraservice.so`, and this lib requires to load `/system/lib` libs, which is possible in Q but does not work anymore in R. 
Is there any way to build a static version of this lib?
